File: plataform/MpOS Plataform/src/br/ufc/mdcc/mpos/net/AbstractServer.java
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package br.ufc.mdcc.mpos.net;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

import org.apache.log4j.Logger;

import br.ufc.mdcc.mpos.net.util.Service;

/**
 * This is a general server implementation.
 * 
 * @author Philipp B. Costa
 */
public abstract class AbstractServer extends Thread {
	protected Logger logger;
	
	private Service service;
	protected String ip;
	
	protected String startMessage;

	public AbstractServer(String ip, Service service, Class<?> cls) {
		super(service.getName());

		this.logger = Logger.getLogger(cls);
		this.service = service;
		this.ip = ip;
	}

	public Service getService() {
		return service;
	}

	protected void close(OutputStream os) {
		try {
			if (os != null) {
				os.close();
			}
		} catch (IOException e) {
			logger.info("IOException was thrown during close Stream");
		}
	}

	protected void close(Socket socket) {
		try {
			if (socket != null) {
				socket.close();
			}
		} catch (IOException e) {
			logger.info("IOException was thrown during close Socket");
		}
	}

	protected void close(InputStream is) {
		try {
			if (is != null) {
				is.close();
			}
		} catch (IOException e) {
			logger.info("IOException was thrown during close Stream");
		}
	}
}

File: plataform/MpOS Plataform Sharp/MpOS/Net/AbstractServer.cs
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using Ufc.MpOS.Net.Util;
using Ufc.MpOS.Util;

namespace Ufc.MpOS.Net
{
	public abstract class AbstractServer
	{
		protected Logger logger;
		private Thread serverThread;
		protected string startMessage;

		protected string ip;

		public Service Service { get; private set; }

		public AbstractServer(string ip, Service service, Type typeCls)
		{
			this.ip = ip;
			logger = Logger.GetLogger(typeCls);
			Service = service;
		}

		public abstract void Run();

		public void Start()
		{
			serverThread = new Thread(Run);
			serverThread.Start();
		}

		protected void Dispose(params IDisposable[] resources)
		{
			if (resources != null)
			{
				foreach (IDisposable res in resources)
				{
					res.Dispose();
				}
			}
		}

		protected MemoryStream ReadStream(NetworkStream stream)
		{
			return ReadStream(new BufferedStream(stream), 1024 * 32);
		}

		//client side on wp didn't has BufferedStream because this, I'm need made
		//the implementation style it...
		protected MemoryStream ReadStream(BufferedStream stream, int bufferSize)
		{
			byte[] streamSize = new byte[4];
			stream.Read(streamSize, 0, streamSize.Length);
			int totalStreamSize = BitConverter.ToInt32(streamSize, 0);

			return ReadStream(stream, bufferSize, totalStreamSize);
		}

		//the client need to sent a 4 byte of stream size!
		protected MemoryStream ReadStream(BufferedStream stream, int bufferSize, int totalStreamSize)
		{
			byte[] buffer = new byte[bufferSize-200];
			MemoryStream memoryStream = new MemoryStream(totalStreamSize);

			//logger.Info(">> RpcTcpServer Receive (totalStreamSize): " + totalStreamSize);
			int totalRead = 0;
			do
			{
				int read = stream.Read(buffer, 0, buffer.Length);
				memoryStream.Write(buffer, 0, read);
				totalRead += read;
				//logger.Info(">> RpcTcpServer Read (totalRead): " + totalRead);
			} while (totalRead < totalStreamSize);

			//start position on memory stream, maybe need discard the header size!
			memoryStream.Position = totalRead - totalStreamSize;
			return memoryStream;
		}

		protected MemoryStream ReadStreamDebug(BufferedStream stream, int bufferSize, int totalStreamSize, ref long totalDownloadTime)
		{
			Stopwatch stopWatch = Stopwatch.StartNew();
			MemoryStream ms = ReadStream(stream, bufferSize, totalStreamSize);
			stopWatch.Stop();

			totalDownloadTime = stopWatch.ElapsedMilliseconds;
			return ms;
		}
	}
}